argilla/src/argilla/datasets/_export/_hub.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2024-present, Argilla, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+from collections import defaultdict
+from tempfile import TemporaryDirectory
+from typing import Any, Type, TYPE_CHECKING, Optional, Union
+from uuid import UUID
+
+from argilla.records import Record
+from argilla.responses import Response
+from argilla._exceptions import NotFoundError
+from argilla._helpers import resolve_hf_datasets_type
+
+if TYPE_CHECKING:
+    from argilla import Dataset
+    from argilla import Argilla
+    from argilla import Workspace
+
+HFDataset = resolve_hf_datasets_type()
+
+
+class HubImportExportMixin:
+    _default_dataset_repo_dir_name = ".argilla"
+    _default_configuration_files = ["settings.json", "dataset.json"]
+
+    def to_hub(
+        self: "Dataset",
+        repo_id: str,
+        generate_card: Optional[bool] = True,
+        *args,
+        **kwargs,
+    ) -> None:
+        """Pushes the `Dataset` to the Hugging Face Hub. If the dataset has been previously pushed to the
+        Hugging Face Hub, it will be updated instead of creating a new dataset repo.
+
+        Parameters:
+            repo_id: the ID of the Hugging Face Hub repo to push the `Dataset` to.
+            generate_card: whether to generate a dataset card for the `FeedbackDataset` in the Hugging Face Hub. Defaults
+                to `True`.
+            *args: the args to pass to `datasets.Dataset.push_to_hub`.
+            **kwargs: the kwargs to pass to `datasets.Dataset.push_to_hub`.
+        """
+
+        from huggingface_hub import HfApi, DatasetCardData
+
+        from argilla.datasets._export.card import (
+            ArgillaDatasetCard,
+            size_categories_parser,
+        )
+
+        hf_api = HfApi()
+
+        hfds = self.records(with_vectors=True, with_responses=True, with_suggestions=True).to_datasets()
+        hfds.push_to_hub(repo_id, *args, **kwargs)
+
+        with TemporaryDirectory() as tmpdirname:
+            self.to_disk(path=tmpdirname, with_records=False)
+            hf_api.upload_folder(
+                folder_path=tmpdirname,
+                repo_id=repo_id,
+                repo_type="dataset",
+                token=kwargs.get("token"),
+                path_in_repo=self._default_dataset_repo_dir_name,
+            )
+
+        if generate_card:
```

Comment:
To reduce the number of commits, you can move the `if generate_card:` section under `with TemporaryDirectory() as tmpdirname:`. Then, instead of `card.push_to_hub(repo_id, repo_type="dataset", token=kwargs.get("token"))`, you can do `card.save(os.path.join(tmpdirname, "README.md"))`. And then you move the `hf_api.upload_folder` call at the end to upload both the dataset and the dataset card in a single commit.

Reply:
**EDIT:** to do the above, you would need to update `self.to_disk` as well:
```py
self.to_disk(path=os.path.join(tmpdirname, self._default_dataset_repo_dir_name), with_records=False)
```

so that the whole `tmpdirname` folder can be uploaded at the root of the repo on the Hub:

```py
hf_api.upload_folder(folder_path=tmpdirname, repo_id=repo_id, repo_type="dataset")
```